pir/pir-internal/src/main/java/com/duckduckgo/pir/internal/common/PirActionsRunner.kt

```diff
@@ -0,0 +1,944 @@
+                ExecuteBrokerAction(
+                    commandsFlow.value.state.copy(
+                        currentActionIndex = commandsFlow.value.state.currentActionIndex + 1,
+                    ),
+                ),
+            )
+        } else {
+            logcat { "PIR-RUNNER (${this@RealPirActionsRunner}): Ignoring $url as next action has been pushed" }
+        }
+    }
+
+    private suspend fun handleCommand(
+        command: Command,
+        continuationResult: Continuation<Result<Unit>>,
+    ) {
+        logcat { "PIR-RUNNER ($this): Handle command: $command" }
+        when (command) {
+            Idle -> {} // Do nothing
+            is HandleBroker -> handleBrokerAction(command.state)
+            is ExecuteBrokerAction -> executeBrokerAction(command.state, command.actionRequestData)
+            is CompleteExecution -> {
+                continuationResult.resume(Result.success(Unit))
+                cleanUpRunner()
+            }
+            // TODO add loading timeout
+            is LoadUrl -> coroutineScope.launch(dispatcherProvider.main()) {
+                detachedWebView!!.loadUrl(command.urlToLoad)
+            }
+
+            is BrokerCompleted -> handleBrokerCompleted(command.state, command.isSuccess)
+            is GetEmail -> handleGetEmail(command.state)
+            is AwaitEmailConfirmation -> handleEmailConfirmation(command.state, command.pollingIntervalSeconds)
+            is GetCaptchaSolution -> handleGetCaptchaSolution(command.state, command.responseData, command.isRetry)
+            is AwaitCaptchaSolution -> handleAwaitCaptchaSolution(command.state, command.pollingIntervalSeconds, command.retries, command.attempt)
+            is SendCaptchaSolution -> handleSendCaptchaSolution(command.state, command.callback)
+        }
+    }
+
+    private fun nextCommand(command: Command) {
+        coroutineScope.launch {
+            commandsFlow.emit(command)
+        }
+    }
+
+    private suspend fun handleSendCaptchaSolution(
+        state: State,
+        callback: String,
+    ) {
+        withContext(dispatcherProvider.main()) {
+            detachedWebView?.evaluateJavascript(callback, null)
+        }
+        nextCommand(
+            ExecuteBrokerAction(
+                state = state.copy(
+                    currentActionIndex = state.currentActionIndex + 1,
+                ),
+            ),
+        )
+    }
+
+    private suspend fun handleAwaitCaptchaSolution(
+        state: State,
+        pollingIntervalSeconds: Int,
+        retries: Int,
+        attempt: Int,
+    ) {
+        val broker = brokersToExecute[state.currentBrokerIndex]
+
+        if (state.transactionID.isEmpty()) {
+            onError(
+                PirErrorReponse(
+                    actionID = broker.actions[state.currentActionIndex].id,
+                    message = "Unable to solve captcha",
+                ),
+            )
+        } else {
+            nativeBrokerActionHandler.pushAction(
```

Comment:
The logic here polls for the solution until we hit max retries (50) every 5 seconds. We can update the frequency here - but so far it does take 2 minutes+ to get the solution